source/Nevermore/ConcurrencyMode.cs

```diff
@@ -0,0 +1,25 @@
+﻿namespace Nevermore
+{
+    public enum ConcurrencyMode
```

Comment:
Food for thought:

```csharp
[Flags]
public enum ConcurrencyMode
{
    Lock = 1
    Log = 1 << 1
}
```

This models the problem better (locking and logging are two different dimensions, and can both move independently), but is probably a worse fit for your code, as you have four implementations of the handler.